Infinite re-sends on Connection Timeout to Database

When rpush loads a notification for sending but fails to save its delivered state due to a connection timeout, it re-sends the same notification until it can reconnect. This results in the user being spammed with the same notification until the timeout situation is resolved.

```
[2014-07-24 17:12:41] [iOS SDK18] 366 sent to a6f61a1c67a7d87d99ce009dc6faad457ad18264f38cebcc350ebb9b901f0f93
[2014-07-24 17:12:41] [iOS SDK18] 367 sent to a6f61a1c67a7d87d99ce009dc6faad457ad18264f38cebcc350ebb9b901f0f93
[2014-07-24 17:12:46] [ERROR] ActiveRecord::ConnectionTimeoutError, could not obtain a database connection within 5.000 seconds (waited 5.000 seconds)
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:190:in `block in wait_poll'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:181:in `loop'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:181:in `wait_poll'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:136:in `block in poll'
/usr/local/rvm/rubies/ruby-2.1.1/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:146:in `synchronize'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:134:in `poll'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:418:in `acquire_connection'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:351:in `block in checkout'
/usr/local/rvm/rubies/ruby-2.1.1/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:350:in `checkout'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:265:in `block in connection'
/usr/local/rvm/rubies/ruby-2.1.1/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:264:in `connection'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/activerecord-4.1.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:294:in `with_connection'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/store/active_record/reconnectable.rb:24:in `with_database_reconnect_and_retry'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/store/active_record.rb:64:in `mark_batch_delivered'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/batch.rb:87:in `complete_delivered'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/batch.rb:75:in `block in complete'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/batch.rb:73:in `each'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/batch.rb:73:in `complete'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/batch.rb:58:in `block in notification_dispatched'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/batch.rb:56:in `synchronize'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/batch.rb:56:in `notification_dispatched'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/dispatcher_loop.rb:49:in `dispatch'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/dispatcher_loop.rb:16:in `block (2 levels) in start'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/dispatcher_loop.rb:15:in `loop'
/var/www/lqd.io/shared/bundle/ruby/2.1.0/gems/rpush-1.0.0/lib/rpush/daemon/dispatcher_loop.rb:15:in `block in start'
[2014-07-24 17:12:46] [ios android sdk6] Connected to feedback.sandbox.push.apple.com:2196
[2014-07-24 17:12:46] [iOS SDK18] Connected to feedback.sandbox.push.apple.com:2196
[2014-07-24 17:12:47] [iOS SDK18] 366 sent to a6f61a1c67a7d87d99ce009dc6faad457ad18264f38cebcc350ebb9b901f0f93
[2014-07-24 17:12:47] [iOS SDK18] 367 sent to a6f61a1c67a7d87d99ce009dc6faad457ad18264f38cebcc350ebb9b901f0f93
```

--- a/lib/rpush/daemon/store/active_record/reconnectable.rb
+++ b/lib/rpush/daemon/store/active_record/reconnectable.rb
@@ -22,7 +22,7 @@ module Rpush
         module Reconnectable
           ADAPTER_ERRORS = [::ActiveRecord::StatementInvalid, PGError, PG::Error,
                             Mysql::Error, Mysql2::Error, ::ActiveRecord::JDBCError,
-                            SQLite3::Exception]
+                            SQLite3::Exception, ::ActiveRecord::ConnectionTimeoutError]
 
           def with_database_reconnect_and_retry
             ::ActiveRecord::Base.connection_pool.with_connection do

--- a/spec/unit/daemon/store/active_record/reconnectable_spec.rb
+++ b/spec/unit/daemon/store/active_record/reconnectable_spec.rb
@@ -43,7 +43,8 @@ describe Rpush::Daemon::Store::ActiveRecord::Reconnectable do
   end
 
   let(:error) { adapter_error_class.new("db down!") }
-  let(:test_double) { TestDouble.new(error, 1) }
+  let(:timeout) { ActiveRecord::ConnectionTimeoutError.new("db lazy!") }
+  let(:test_doubles) { [TestDouble.new(error, 1), TestDouble.new(timeout, 1)]  }
 
   before do
     @logger = double("Logger", info: nil, error: nil, warn: nil)
@@ -51,37 +52,37 @@ describe Rpush::Daemon::Store::ActiveRecord::Reconnectable do
 
     ActiveRecord::Base.stub(:clear_all_connections!)
     ActiveRecord::Base.stub(:establish_connection)
-    test_double.stub(:sleep)
+    test_doubles.each { |td| allow(td).to receive(:sleep) }
   end
 
   it "should log the error raised" do
     Rpush.logger.should_receive(:error).with(error)
-    test_double.perform
+    test_doubles.each(&:perform)
   end
 
   it "should log that the database is being reconnected" do
     Rpush.logger.should_receive(:warn).with("Lost connection to database, reconnecting...")
-    test_double.perform
+    test_doubles.each(&:perform)
   end
 
   it "should log the reconnection attempt" do
     Rpush.logger.should_receive(:warn).with("Attempt 1")
-    test_double.perform
+    test_doubles.each(&:perform)
   end
 
   it "should clear all connections" do
     ActiveRecord::Base.should_receive(:clear_all_connections!)
-    test_double.perform
+    test_doubles.each(&:perform)
   end
 
   it "should establish a new connection" do
     ActiveRecord::Base.should_receive(:establish_connection)
-    test_double.perform
+    test_doubles.each(&:perform)
   end
 
   it "should test out the new connection by performing a count" do
-    Rpush::Client::ActiveRecord::Notification.should_receive(:count)
-    test_double.perform
+    Rpush::Client::ActiveRecord::Notification.should_receive(:count).twice
+    test_doubles.each(&:perform)
   end
 
   context "when the reconnection attempt is not successful" do
@@ -97,19 +98,38 @@ describe Rpush::Daemon::Store::ActiveRecord::Reconnectable do
       Rpush::Client::ActiveRecord::Notification.instance_variable_set("@error", error)
     end
 
-    it "should log the 2nd attempt" do
-      Rpush.logger.should_receive(:warn).with("Attempt 2")
-      test_double.perform
+    describe "error behaviour" do
+      it "should log the 2nd attempt" do
+        Rpush.logger.should_receive(:warn).with("Attempt 2")
+        test_doubles[0].perform
+      end
+
+      it "should log errors raised when the reconnection is not successful" do
+        Rpush.logger.should_receive(:error).with(error)
+        test_doubles[0].perform
+      end
+
+      it "should sleep to avoid thrashing when the database is down" do
+        expect(test_doubles[0]).to receive(:sleep).with(2)
+        test_doubles[0].perform
+      end
     end
 
-    it "should log errors raised when the reconnection is not successful" do
-      Rpush.logger.should_receive(:error).with(error)
-      test_double.perform
-    end
+    describe "timeout behaviour" do
+      it "should log the 2nd attempt" do
+        Rpush.logger.should_receive(:warn).with("Attempt 2")
+        test_doubles[1].perform
+      end
 
-    it "should sleep to avoid thrashing when the database is down" do
-      test_double.should_receive(:sleep).with(2)
-      test_double.perform
+      it "should log errors raised when the reconnection is not successful" do
+        Rpush.logger.should_receive(:error).with(error)
+        test_doubles[1].perform
+      end
+
+      it "should sleep to avoid thrashing when the database is down" do
+        expect(test_doubles[1]).to receive(:sleep).with(2)
+        test_doubles[1].perform
+      end
     end
   end
 end

